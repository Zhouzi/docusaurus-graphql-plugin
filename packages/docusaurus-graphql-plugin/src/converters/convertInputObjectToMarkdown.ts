import { GraphQLInputObjectType } from "graphql";
import { MarkdownConverterOptions } from "../types";
import { pushArguments } from "./pushArguments";

export function convertInputObjectToMarkdown(
  inputObject: GraphQLInputObjectType,
  { getTypePath }: MarkdownConverterOptions
): string {
  const lines: string[] = [];

  lines.push(`## ${inputObject.name}`, `\n\n`);
  lines.push(inputObject.description || "", `\n\n`);

  pushArguments(lines, Object.values(inputObject.getFields()), {
    getTypePath,
  });

  return lines.join("");
}
