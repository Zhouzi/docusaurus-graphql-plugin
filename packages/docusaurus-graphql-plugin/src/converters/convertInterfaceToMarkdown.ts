import { GraphQLInterfaceType, GraphQLObjectType } from "graphql";
import { MarkdownConverterOptions } from "../types";
import { pushFields } from "./pushFields";
import { pushInterfaces } from "./pushInterfaces";

export function convertInterfaceToMarkdown(
  inter: GraphQLInterfaceType,
  implementedBy: Array<GraphQLObjectType | GraphQLInterfaceType>,
  { getTypePath }: MarkdownConverterOptions
): string {
  const lines: string[] = [];

  lines.push(`## ${inter.name}`, `\n\n`);
  lines.push(inter.description || "", `\n\n`);

  const subInterfaces = inter.getInterfaces();
  if (subInterfaces.length > 0) {
    pushInterfaces(lines, subInterfaces, { getTypePath });
  }

  if (implementedBy.length > 0) {
    lines.push(
      `<p style={{ marginBottom: "0.4em" }}><strong>Implemented by</strong></p>`,
      `\n\n`
    );
    implementedBy.forEach((object) => {
      lines.push(`- [${object.name}](${getTypePath(object)})`, `\n`);
    });
    lines.push(`\n`);
  }

  const fields = Object.values(inter.getFields());
  pushFields(lines, fields, { getTypePath });

  return lines.join("");
}
