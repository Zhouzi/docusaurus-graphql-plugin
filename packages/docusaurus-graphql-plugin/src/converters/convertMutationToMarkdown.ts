import { GraphQLField } from "graphql";
import { MarkdownConverterOptions } from "../types";
import { pushArguments } from "./pushArguments";

export function convertMutationToMarkdown(
  mutation: GraphQLField<any, any>,
  { getTypePath }: MarkdownConverterOptions
): string {
  const lines: string[] = [];

  lines.push(`## ${mutation.name}`, `\n\n`);
  lines.push(
    `**Type:** [${mutation.type.toJSON()}](${getTypePath(mutation.type)})`,
    `\n\n`
  );

  if (mutation.deprecationReason) {
    lines.push(`> Deprecated: ${mutation.deprecationReason}`, `\n\n`);
  }

  lines.push(mutation.description || "", `\n\n`);

  if (mutation.args.length > 0) {
    pushArguments(lines, mutation.args, { getTypePath });
  }

  return lines.join("");
}
