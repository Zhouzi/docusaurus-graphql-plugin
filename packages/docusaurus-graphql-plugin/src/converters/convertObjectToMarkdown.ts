import { GraphQLObjectType } from "graphql";
import { MarkdownConverterOptions } from "../types";
import { pushFields } from "./pushFields";
import { pushInterfaces } from "./pushInterfaces";

export function convertObjectToMarkdown(
  object: GraphQLObjectType,
  { getTypePath }: MarkdownConverterOptions
): string {
  const lines: string[] = [];

  lines.push(`## ${object.name}`, `\n\n`);
  lines.push(object.description || "", `\n\n`);

  const interfaces = object.getInterfaces();
  if (interfaces.length > 0) {
    pushInterfaces(lines, interfaces, { getTypePath });
  }

  const fields = Object.values(object.getFields());
  pushFields(lines, fields, { getTypePath });

  return lines.join("");
}
