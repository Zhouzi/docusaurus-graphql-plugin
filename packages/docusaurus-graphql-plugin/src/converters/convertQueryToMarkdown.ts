import { GraphQLField } from "graphql";
import { MarkdownConverterOptions } from "../types";
import { pushArguments } from "./pushArguments";

export function convertQueryToMarkdown(
  query: GraphQLField<any, any>,
  { getTypePath }: MarkdownConverterOptions
): string {
  const lines: string[] = [];

  lines.push(`## ${query.name}`, `\n\n`);
  lines.push(
    `**Type:** [${query.type.toJSON()}](${getTypePath(query.type)})`,
    `\n\n`
  );

  if (query.deprecationReason) {
    lines.push(`> Deprecated: ${query.deprecationReason}`, `\n\n`);
  }

  lines.push(query.description || "", `\n\n`);

  if (query.args.length > 0) {
    pushArguments(lines, query.args, { getTypePath });
  }

  return lines.join("");
}
