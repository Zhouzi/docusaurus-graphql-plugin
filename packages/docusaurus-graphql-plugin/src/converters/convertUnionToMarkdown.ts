import { GraphQLUnionType } from "graphql";
import { MarkdownConverterOptions } from "../types";

export function convertUnionToMarkdown(
  union: GraphQLUnionType,
  { getTypePath }: MarkdownConverterOptions
): string {
  const lines: string[] = [];

  lines.push(`## ${union.name}`, `\n\n`);
  lines.push(union.description || "", `\n\n`);

  lines.push(
    `<p style={{ marginBottom: "0.4em" }}><strong>Possible types</strong></p>`,
    `\n\n`
  );
  union.getTypes().forEach((type) => {
    lines.push(`- [${type.name}](${getTypePath(type)})`, `\n`);
  });
  lines.push(`\n`);

  return lines.join("");
}
