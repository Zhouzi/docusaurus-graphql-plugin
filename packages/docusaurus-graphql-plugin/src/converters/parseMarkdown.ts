import marked from "marked";

export function parseMarkdown(markdown: string): string {
  const walkTokens = (token: marked.Token) => {
    if (token.type === "text") {
      // make the Markdown compatible with MDX by escaping curly braces
      token.text = token.text
        .replace(/\{/g, "&lbrace;")
        .replace(/\}/g, "&rbrace;");
    }
  };
  return marked
    .parse(markdown, {
      // marked's types are buggy
      walkTokens: walkTokens as any,
    })
    .trim();
}
