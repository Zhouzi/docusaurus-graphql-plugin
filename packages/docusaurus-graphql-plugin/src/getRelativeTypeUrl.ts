import {
  GraphQLNamedType,
  GraphQLType,
  isListType,
  isNonNullType,
} from "graphql";
import { Slugger } from "marked";
import { convertersList } from "./converters";

const slugger = new Slugger();
const sluggify = (name: string) => slugger.slug(name, { dryrun: true });

function getBaseType(type: GraphQLType): GraphQLNamedType {
  if (isNonNullType(type)) {
    return getBaseType(type.ofType);
  }

  if (isListType(type)) {
    return getBaseType(type.ofType);
  }

  return type;
}

export function getRelativeTypeUrl(type: GraphQLType): string {
  const baseType = getBaseType(type);
  const converter = convertersList.find((otherConverter) =>
    otherConverter.matches(baseType)
  );

  if (converter == null) {
    throw new Error(`Can't generate an URL for type "${baseType.name}"`);
  }

  return `/${converter.id}#${sluggify(baseType.name)}`;
}
