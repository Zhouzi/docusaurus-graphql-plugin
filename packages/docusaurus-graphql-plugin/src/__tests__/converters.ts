import { buildSchema } from "graphql";
import { getRelativeTypeUrl } from "../getRelativeTypeUrl";
import * as converters from "../converters";

const markdownConverterOptions = {
  getTypePath: getRelativeTypeUrl,
};

describe("converters", () => {
  describe("queries", () => {
    it("should convert queries to markdown", () => {
      const schema = buildSchema(/* GraphQL */ `
        type Query {
          """
          Query to get a user.
          """
          user("The user's id." id: ID!): User!

          """
          Query to get an admin.
          """
          admin("The admin's id." id: ID!): User!
            @deprecated(
              reason: "admin will be removed with the next major release, use { user } instead"
            )
        }
        type User {
          name: String!
        }
      `);
      expect(
        converters.queries.convertToMarkdown(schema, markdownConverterOptions)
      ).toMatchSnapshot();
    });

    it("should support query referencing Query", () => {
      const schema = buildSchema(/* GraphQL */ `
        type Query {
          """
          Hack to workaround https://github.com/facebook/relay/issues/112 re-exposing the root query object.
          """
          relay: Query!
        }
      `);
      expect(
        converters.queries.convertToMarkdown(schema, markdownConverterOptions)
      ).toMatchSnapshot();
    });
  });

  describe("mutations", () => {
    it("should convert mutations to markdown", () => {
      const schema = buildSchema(/* GraphQL */ `
        type Mutation {
          """
          Mutation to create a new user.
          """
          createUser("The new user's name" name: String!): User!

          """
          Mutation to create a new admin.
          """
          createAdmin("The new admin's name" name: String!): User!
            @deprecated(
              reason: "{ createAdmin } will be removed with the next major release"
            )
        }
        type User {
          name: String!
        }
      `);
      expect(
        converters.mutations.convertToMarkdown(schema, markdownConverterOptions)
      ).toMatchSnapshot();
    });
  });

  describe("objects", () => {
    it("should convert objects to markdown", () => {
      const schema = buildSchema(/* GraphQL */ `
        type User {
          """
          This field's description contains special characters that should be escaped, such as { and }.

          It should also work within inline code such as \`{ example }\`
          """
          id: ID!
        }
      `);
      expect(
        converters.objects.convertToMarkdown(schema, markdownConverterOptions)
      ).toMatchSnapshot();
    });

    it("should list the objects' interfaces", () => {
      const schema = buildSchema(/* GraphQL */ `
        interface User {
          name: String!
        }

        """
        A user with special rights.
        """
        type Admin implements User {
          "The admin's name."
          name: String!
        }
      `);
      expect(
        converters.objects.convertToMarkdown(schema, markdownConverterOptions)
      ).toMatchSnapshot();
    });
  });

  describe("interfaces", () => {
    it("should convert interfaces to markdown", () => {
      const schema = buildSchema(/* GraphQL */ `
        """
        Properties that are common to the different kind of users.
        """
        interface User {
          "The user's name."
          name: String!
        }
      `);
      expect(
        converters.interfaces.convertToMarkdown(
          schema,
          markdownConverterOptions
        )
      ).toMatchSnapshot();
    });

    it("should list the interface's interfaces", () => {
      const schema = buildSchema(/* GraphQL */ `
        """
        Properties that are common to the different kind of users.
        """
        interface User {
          "The user's name."
          name: String!
        }

        """
        A special kind of user.
        """
        interface Admin implements User {
          "The admin's special power."
          power: String!
        }
      `);
      expect(
        converters.interfaces.convertToMarkdown(
          schema,
          markdownConverterOptions
        )
      ).toMatchSnapshot();
    });

    it("should list the interface's implementations", () => {
      const schema = buildSchema(/* GraphQL */ `
        """
        Properties that are common to the different kind of users.
        """
        interface User {
          "The user's name."
          name: String!
        }

        """
        A special kind of user.
        """
        type Admin implements User {
          "The admin's name."
          name: String!
        }
      `);
      expect(
        converters.interfaces.convertToMarkdown(
          schema,
          markdownConverterOptions
        )
      ).toMatchSnapshot();
    });
  });

  describe("enums", () => {
    it("should convert enums to markdown", () => {
      const schema = buildSchema(/* GraphQL */ `
        """
        Represents a user's role.
        """
        enum UserRole {
          "User has admin rights."
          Admin
          "User has no particular rights."
          User
          "Unknown user"
          Anonymous
            @deprecated(
              reason: "{ Anonymous } will be removed with the next major release"
            )
        }
      `);
      expect(
        converters.enums.convertToMarkdown(schema, markdownConverterOptions)
      ).toMatchSnapshot();
    });
  });

  describe("unions", () => {
    it("should convert unions to markdown", () => {
      const schema = buildSchema(/* GraphQL */ `
        type User {
          name: String!
        }
        type Droid {
          name: String!
        }

        """
        Combination of the possible humanoids.
        """
        union Humanoid = User | Droid
      `);
      expect(
        converters.unions.convertToMarkdown(schema, markdownConverterOptions)
      ).toMatchSnapshot();
    });
  });

  describe("inputObjects", () => {
    it("should convert inputObjects to markdown", () => {
      const schema = buildSchema(/* GraphQL */ `
        """
        Input to create a new user.
        """
        input CreateUserInput {
          "The new user's name."
          name: String!
        }
      `);
      expect(
        converters.inputObjects.convertToMarkdown(
          schema,
          markdownConverterOptions
        )
      ).toMatchSnapshot();
    });
  });

  describe("scalars", () => {
    it("should convert scalars to markdown", () => {
      const schema = buildSchema(/* GraphQL */ `
        """
        A ISO date time.
        """
        scalar DateTime
      `);
      expect(
        converters.scalars.convertToMarkdown(schema, markdownConverterOptions)
      ).toMatchSnapshot();
    });
  });
});
