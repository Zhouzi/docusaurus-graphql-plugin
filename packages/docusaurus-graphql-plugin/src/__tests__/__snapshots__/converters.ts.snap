// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`converters enums should convert enums to markdown 1`] = `
"## UserRole

Represents a user's role.

<p style={{ marginBottom: \\"0.4em\\" }}><strong>Values</strong></p>

<table>
<thead><tr><th>Value</th><th>Description</th></tr></thead>
<tbody>
<tr>
<td>Admin</td>
<td>
<p>User has admin rights.</p>
</td>
</tr>
<tr>
<td>User</td>
<td>
<p>User has no particular rights.</p>
</td>
</tr>
<tr>
<td>Anonymous</td>
<td>
<blockquote>Deprecated: &#123; Anonymous &#125; will be removed with the next major release</blockquote>
<p>Unknown user</p>
</td>
</tr>
</tbody>
</table>

"
`;

exports[`converters inputObjects should convert inputObjects to markdown 1`] = `
"## CreateUserInput

Input to create a new user.

<p style={{ marginBottom: \\"0.4em\\" }}><strong>Arguments</strong></p>

<table>
<thead><tr><th>Name</th><th>Description</th></tr></thead>
<tbody>
<tr>
<td>
name<br />
<a href=\\"/scalars#string\\"><code>String!</code></a>
</td>
<td>
<p>The new user&#39;s name.</p>
</td>
</tr>
</tbody>
</table>

"
`;

exports[`converters interfaces should convert interfaces to markdown 1`] = `
"## User

Properties that are common to the different kind of users.

<p style={{ marginBottom: \\"0.4em\\" }}><strong>Fields</strong></p>

<table>
<thead><tr><th>Name</th><th>Description</th></tr></thead>
<tbody>
<tr>
<td>
name<br />
<a href=\\"/scalars#string\\"><code>String!</code></a>
</td>
<td>
<p>The user&#39;s name.</p>
</td>
</tr>
</tbody>
</table>

"
`;

exports[`converters interfaces should list the interface's implementations 1`] = `
"## User

Properties that are common to the different kind of users.

<p style={{ marginBottom: \\"0.4em\\" }}><strong>Implemented by</strong></p>

- [Admin](/objects#admin)

<p style={{ marginBottom: \\"0.4em\\" }}><strong>Fields</strong></p>

<table>
<thead><tr><th>Name</th><th>Description</th></tr></thead>
<tbody>
<tr>
<td>
name<br />
<a href=\\"/scalars#string\\"><code>String!</code></a>
</td>
<td>
<p>The user&#39;s name.</p>
</td>
</tr>
</tbody>
</table>

"
`;

exports[`converters interfaces should list the interface's interfaces 1`] = `
"## Admin

A special kind of user.

<p style={{ marginBottom: \\"0.4em\\" }}><strong>Implements</strong></p>

- [User](/interfaces#user)

<p style={{ marginBottom: \\"0.4em\\" }}><strong>Fields</strong></p>

<table>
<thead><tr><th>Name</th><th>Description</th></tr></thead>
<tbody>
<tr>
<td>
power<br />
<a href=\\"/scalars#string\\"><code>String!</code></a>
</td>
<td>
<p>The admin&#39;s special power.</p>
</td>
</tr>
</tbody>
</table>

## User

Properties that are common to the different kind of users.

<p style={{ marginBottom: \\"0.4em\\" }}><strong>Implemented by</strong></p>

- [Admin](/interfaces#admin)

<p style={{ marginBottom: \\"0.4em\\" }}><strong>Fields</strong></p>

<table>
<thead><tr><th>Name</th><th>Description</th></tr></thead>
<tbody>
<tr>
<td>
name<br />
<a href=\\"/scalars#string\\"><code>String!</code></a>
</td>
<td>
<p>The user&#39;s name.</p>
</td>
</tr>
</tbody>
</table>

"
`;

exports[`converters mutations should convert mutations to markdown 1`] = `
"## createAdmin

**Type:** [User!](/objects#user)

> Deprecated: &#123; createAdmin &#125; will be removed with the next major release

Mutation to create a new admin.

<p style={{ marginBottom: \\"0.4em\\" }}><strong>Arguments</strong></p>

<table>
<thead><tr><th>Name</th><th>Description</th></tr></thead>
<tbody>
<tr>
<td>
name<br />
<a href=\\"/scalars#string\\"><code>String!</code></a>
</td>
<td>
<p>The new admin&#39;s name</p>
</td>
</tr>
</tbody>
</table>

## createUser

**Type:** [User!](/objects#user)

Mutation to create a new user.

<p style={{ marginBottom: \\"0.4em\\" }}><strong>Arguments</strong></p>

<table>
<thead><tr><th>Name</th><th>Description</th></tr></thead>
<tbody>
<tr>
<td>
name<br />
<a href=\\"/scalars#string\\"><code>String!</code></a>
</td>
<td>
<p>The new user&#39;s name</p>
</td>
</tr>
</tbody>
</table>

"
`;

exports[`converters objects should convert objects to markdown 1`] = `
"## User



<p style={{ marginBottom: \\"0.4em\\" }}><strong>Fields</strong></p>

<table>
<thead><tr><th>Name</th><th>Description</th></tr></thead>
<tbody>
<tr>
<td>
id<br />
<a href=\\"/scalars#id\\"><code>ID!</code></a>
</td>
<td>
<p>This field&#39;s description contains special characters that should be escaped, such as &#123; and &#125;.</p>
<p>It should also work within inline code such as <code>&#123; example &#125;</code></p>
</td>
</tr>
</tbody>
</table>

"
`;

exports[`converters objects should list the objects' interfaces 1`] = `
"## Admin

A user with special rights.

<p style={{ marginBottom: \\"0.4em\\" }}><strong>Implements</strong></p>

- [User](/interfaces#user)

<p style={{ marginBottom: \\"0.4em\\" }}><strong>Fields</strong></p>

<table>
<thead><tr><th>Name</th><th>Description</th></tr></thead>
<tbody>
<tr>
<td>
name<br />
<a href=\\"/scalars#string\\"><code>String!</code></a>
</td>
<td>
<p>The admin&#39;s name.</p>
</td>
</tr>
</tbody>
</table>

"
`;

exports[`converters queries should convert queries to markdown 1`] = `
"## admin

**Type:** [User!](/objects#user)

> Deprecated: admin will be removed with the next major release, use &#123; user &#125; instead

Query to get an admin.

<p style={{ marginBottom: \\"0.4em\\" }}><strong>Arguments</strong></p>

<table>
<thead><tr><th>Name</th><th>Description</th></tr></thead>
<tbody>
<tr>
<td>
id<br />
<a href=\\"/scalars#id\\"><code>ID!</code></a>
</td>
<td>
<p>The admin&#39;s id.</p>
</td>
</tr>
</tbody>
</table>

## user

**Type:** [User!](/objects#user)

Query to get a user.

<p style={{ marginBottom: \\"0.4em\\" }}><strong>Arguments</strong></p>

<table>
<thead><tr><th>Name</th><th>Description</th></tr></thead>
<tbody>
<tr>
<td>
id<br />
<a href=\\"/scalars#id\\"><code>ID!</code></a>
</td>
<td>
<p>The user&#39;s id.</p>
</td>
</tr>
</tbody>
</table>

"
`;

exports[`converters queries should support query referencing Query 1`] = `
"## relay

**Type:** [Query!](/queries#query)

Hack to workaround https://github.com/facebook/relay/issues/112 re-exposing the root query object.

"
`;

exports[`converters scalars should convert scalars to markdown 1`] = `
"## Boolean

The \`Boolean\` scalar type represents \`true\` or \`false\`.

## DateTime

A ISO date time.

## String

The \`String\` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.

"
`;

exports[`converters unions should convert unions to markdown 1`] = `
"## Humanoid

Combination of the possible humanoids.

<p style={{ marginBottom: \\"0.4em\\" }}><strong>Possible types</strong></p>

- [User](/objects#user)
- [Droid](/objects#droid)

"
`;
